### PR TITLE
Mackie control

### PR DESCRIPTION
**Pull request by dehnhardt**
Dear Rui,
a few weeks ago I found a remote control App for DAWs (http://www.humatic.de/htools/touchdaw/), which fits my needs to possibly remote control some of the functions of qtractor. It uses the "Mackie Control" (see Page 240 in https://people.ok.ubc.ca/creative/MacManuals/Logic7_DedicatedCntrlSurfaceInfo.pdf) MIDI implementation. Unfortunately the controllers section of qtractor was not powerful enough to emulate this protocol. I fiddled around with mididings but that was not successful. So I decided to extend the controllers section in qtractor. In this pull request you can find my modifications:
* enable an upper limit for the parameters (the protocol uses note on events for solo, mute, rec, so I had to add several lines with note on events for the different buttuns)
* enable different control modes. The buttons send from the controller are push buttons (sending "note on  127" followed by "note on 0" independent from the state)
* enable encoder mode (pan control sends one value to increase the pan value and another one to decrease the totel value)
* enable to feedback all events to the controller (which was already prepared in qtractor but not enabled)

What currently works (only Channels 1-8 are supported in the Protocol) :
* Buttons:  Rec, Mute, Solo, Monitor with feedback to enable/disable the LEDs
* Fader: Midi and Audio but: The audio faders us gain values fron 0.0f to 2.0f so I implemented a hack hack hack! to double the value whe receiving an event and to bisect it when sending the event to the controller. I have cemmented this parts.
* Pan: recieving the events from the controller works, feedback the changes to the controller does not work, as the events recieved from the controller are completely different from the events send, which is currently not possible in qtractor.

I tried to copy your style as much as possible, but I'm sure you wil still find some errors - but I woud be glad if you can review the changes and give some feedback. BTW: I'm everything but a UI designer, so please don't blame me for the gui;-)

Holger

PS: Find a controller setting file here: https://cloud.dehnhardt.org/index.php/s/fjgn141s92kPgj7

### CODE REVIEW

**Comment by rncbc** [comment]
Hi Holger,

May you care to explain what this "parameter limit" purpose and function? And that kind of switch vs. push button modes? Shouldn't it be more like "latch" vs. "momentary" mode that is only applicable to toggling subjects eg. track_monitor, track_record, track_mute and track_solo ? Yes this is something that is missing on these generic controllers but not on the discrete ones (cf. "Latch" setting),

Overall, I tend to find this PR quite an over-complication: it sounds like you went through a lot to support one very specific control protocol idiosyncrasies of (old) Mackie control surfaces. I believe it would be a lot easier making TouchDAW to use simpler/straightforward MIDI controls or configuration--yes, I've done it once, years ago though, and it worked a treat, IIRC.

All that to say, and don't take this as _ad hominem_ please, that I don't really buy the need to support that bizarre "encoder" thing, which as you say is still not working anyway. I'd rather invest on bringing up a non-latched/momentary toggle support into the scene. *wink* *wink* ;)

hth.
cheers

**Comment by dehnhardt** [comment]
Dear Rui,
wow, that was a really quick reply!
the Mackie protocol still seems to be the "smallest common denominator" (I don't know if the saying is also used in english) as nearly every major DAW supports this protocol and during my search for the protocol specification I found, that a lot of controllers suppert this protocol as well.
But my aim was not to just implement that protocol but to enable such protocols to be configured. (Might be a long way to do so...)
First two screenshots: 
image
image

So I try to explain in more detail:
**Parameter limit:**
In that protocol, note on events are used for more than one Command: record, mute, solo and edit switches. 
```
    Note on events           Buttons
    (Note Numbers)
     0  1  2  3  4  5  6  7  record switchs channel 1-8 
     8  9 10 11 12 13 14 15  solo switches channel 1-8
    16 17 18 19 20 21 22 23  mute switches
```
This could possibly be used with other events than note on/off
I first had an implementation which looked up the controller configuration with the smallest note number as a parameter which is gerater than the recieved note number, but since the controller data sets are not sorted, all lines had to be searched through, which I thought was too time consuming.
**Latch vs. momentary:** Uuuh, I've search for a proper translation for the german words 'Schalter' (latch) and 'Taster' (momentary) an only fount 'push button' and 'switch'. But your latch and momentary are by far the better translations! So for better understanding of what I've done we can replace "push button" with "momentary" and "switch button" with "latch".
In the command mode dropdown you can choose between latch and momentary if you configure one of the buttons (rec, solo, mute) Is this what you meant with

> I'd rather invest on bringing up a non-latched/momentary toggle support into the scene. wink wink ;)

**Encoder:** Read as 'rotary encoder':  Rotary encoders only emit the direction (clockwise or counter clockwise) and sometimes a delta. When the track panning command is selected, you can choose between "value", which asssumes, the value of the pan pot is send by the controller  (this is the behaviour as it has been before) and "encoder" which currently looks, if the sended value is less or greater than the [f|i]midScale and ads/substracts a (currently) fixed value to the current value of the pan pot if none of the outer bounds are already reached.

Holger

**Comment by rncbc** [comment]
Hi Holger, thanks for you thorough explanations. Have here some more follow-up comments:

re. **Parameter limit**: shouldn't this better called semantically-wise an **Offset limit** or ""maximum" or else, as it applies to only when _Track + offset_ is in effect?

re. **Latch", "Momentary** : yes, this applies to toggle subjects (monitor, rec, mute, solo); the former is the current behavior and should stay as the default one, while the later is now being proposed to be implemented.

re. **Value**, **Delta** or Encoder : apply to numeric or scalar subjects (gain, volume, panning as in sliders, faders or rotary knobs) being the first the current default mode of operation while the second is about to be supported as new.

I believe your/this PR should be refactored quite a bit to be acceptable though. To that effect will setup an upstream branch named after something like "midicontrolx". I'll also do all the refactoring from your nice base work here. Holy thanks for that, you just give me something to work on this dull winter week-end  ;)

Seeya

**Comment by dehnhardt** [comment]
Hi Rui, glad I wasn't that wrong with my ideas - and of course I expected that some refactoring is neccessarry. Digging in such a big codebase is always some kind of "don't know what I'm doing here";-)
May I suggest one more thing: As I wrote in my first comment, using the encoder does work from the controller to qtractor but not in the other direction, because the expected MIDI events are different. (Which seems to be common. I looked into the HUI specifications and there are some of such occurences) 
It might be a good idea, to add a direction flag (1 = from controller to qtractor, 2 = from qtractor to controller and 3 for.... yes you are right: both directions;-) to each line so it is possible to distinguish between incoming and outgoing events?
Happy coding 
Holger

**Comment by dehnhardt** [comment]
BTW: ... I can add this change too before you start refactoring. I didn't want to leave anything unfinished, but before I continue I wanted to make sure you agree with this kind of changes.

**Comment by rncbc** [comment]
hi, the refactoring base branch is already upstream: [midicontrolx](https://github.com/rncbc/qtractor/tree/midicontrolx).
Basically I laid out the infrastructure to support your **Paramer limit** (cf `trackOffset()`, `trackLimit()`) and the `Delta` value mode of operation that flags for **Momentary** mode on toggling subjects (monitor, record, mute, solo) and **Encoder** for scalar ones (gain, panning).
Still on the drawing bench; later today I'll start adapting the UI and later the actual operational/functional stuff.
Stay tuned. 
Seeya

**Comment by rncbc** [comment]
hi, update to the refactoring [midicontrolx](https://github.com/rncbc/qtractor/tree/midicontrolx) branch:
- the UI (View/Controllers...) is now fully functional; note that the .qtc file format changed a bit, keeping full backward compatibility though.

The converted version for your file maybe found here: [MackieMappingsSmall.qtc](http://www.rncbc.org/datahub/MackieMappingsSmall.qtc).

Still missing the real operational stuff, using the new possible predicates (`qtractorMidiControl::MapVal::trackOffset()`, `trackLimit()` and `isDelta()`of course).

Byee

**Comment by rncbc** [comment]
hi Holger,

One last commit for today: https://github.com/rncbc/qtractor/commit/b8ff2fa8 :-) so called **Delta** flag (aka. _Momentary_ for toggled and _Encoded_ for scalar subjects resp.) is now in place!

Refer to the following translation map:
- toggled subjects (monitor, record, solo, mute): Delta = Momentary, same as your/former "push_button" mode;
- scalar subjects (gain, panning): Delta same as your/former "encoded" mode;

However though, I can't decide on how to implement **Feedback** on these corner cases--the (old) way is still in effect, that is, subject state values are still fed back exactly as before ;)

Please try, test && tell :)
Cheers.

**Comment by dehnhardt** [comment]
Hi Rui,
first my current configuration (did it by hand to thest the configuration as well...):
image
And here are my first results: 
* When using the buttons in qtractor, the buttons in the external controller are set as expected. 
* When using the buttons in the external controller, only the monitor-buttons are toggled, the other buttons simply do not change their status. 
  * Note: The configuration for the monitor buttons is the first section in the .qtc file. If I change the order in the qtc. file so that the solo buttons are the first section, I can use them to switch the buttons in qtractor. But even this behaviou is not consistent. I have to debug it later.
  * When switching the buttons in the external controller, nothing is fed back to the controller. It should be:
```
[rec button channel 1 is off]
c: controller 
q: qtractor
c: Note on 0 127  ->    q: switch on rec button
c: Note on 0 0    ->    q: ignore
q: Note on 0 127  ->    c: switch on record led
Now the record status is on in qtractor and controller

c: Note on 0 127  ->    q: switch off rec button
c: Note on 0 0    ->    q: ignore
q: Note on 0 0    ->    c: switch off record led
Now the record status is off in qtractor and controller
```
In short: The current status of the button must be fed back to the controller.

* Fader: They work in both directions. If I move the faders in the controller, the faders in qtractor are moved as well and vice versa.
  * But: when moving faders in an audio channel more than 0db, the faders in the controller start from zero again and consequential: When moving the faders in the controller to the top, the audio faders in qtractor are only moved to 0db. There should be a different translation for the audio faders, as their value range is 0.0 to 2.0 in opposite to the midi faders which only have a range from 0.0 to 1.0.

* Pan Pots: Whatever I do in the controller, the pan pots in qtractor are moved to the leftmost position and stay there. 
  * Feedback for the pan pots from qtractor to the controller does not wors - but that is expected, as the feedback is completely different from the values which are send. (This did not work with in my changes too - as I wrote before)

I don't know if I find time this evening to dig deeper into the problems. (My family...;-) If not, I will do it tomorrow.

All the best
Holger

**Comment by dehnhardt** [comment]
One first finding:
in my commit https://github.com/rncbc/qtractor/pull/118/commits/468d11006d99e9d039d281e67e8d996443514f3d
I enabled the feedback for the commands when the feedback flag is set. This resolves the problems with the button feedback. (2.i)

**Comment by dehnhardt** [comment]
One other finding (to resolve 2.i):
the MapKey.match method must ensure that 
```
param >= (param() & TrackParamMask)
and
param <= ((param() & TrackParamMask) + limit)
```

**Comment by rncbc** [comment]
Hi,

re. 2.i := your last finding was absolutely awesome! got rid of the `MapKey::match()`method and moved the matching criteria into `qtractorMidiControl::findEvent()` where it was its only caller before. -- https://github.com/rncbc/qtractor/commit/af0d0c91 might have fixed it.

re. 2.ii :- well, this is to stay as is: in principle, "feedback" should only to occur when the controlled subject is changed at its end-point initiative ie. qtractor; if the change is initiated on the external controller then there should be no feedback as direct causality. So I think everything is working as designed.

re. Faders. i :- this (dang old) issue has to be addressed, yes but it should be tackled at the audio track gain subject/observer side, not in the generic MIDI controllers side; the fact is that the mapped values should only apply on the -70dB (>0.0f) to 0dBfs (=1.0f) interval--anything above 0dB (>1.0f) should be capped.

hth.
cheers

**Comment by dehnhardt** [comment]
2.i: Will try out tomorrow

> re. 2.ii :- well, this is to stay as is: in principle, "feedback" should only to occur when the controlled subject is changed at its end-point initiative ie. qtractor; if the change is initiated on the external controller then there should be no feedback as direct causality. So I think everything is working as designed.

As far as I can tell, this would rule out the correct use of almost all hardware controllers. Normally - after sending the event - they wait for feedback from the DAW and set the status (usually by switching an LED on and off) depending on the status of the DAW. This ensures that the two status values are synchronized.

**Comment by rncbc** [comment]
re. 2.ii ...

> As far as I can tell, this would rule out the correct use of almost all hardware controllers. Normally - after sending the event - they wait for feedback from the DAW and set the status (usually by switching an LED on and off) depending on the status of the DAW. This ensures that the two status values are synchronized.

True, but that is about a conversational protocol application, and not the purpose at all of this generic MIDI controllers assignment feature--nor even of the discrete ones--and never were.

The "feedback" option is only for the cases where the external controller recognizes and accepts the state of one specific element in particular with very same function and precise, byte for byte, MIDI sequence it sends to the controlled party (qtractor) in the first place.

In short, "feedback" is not the same nor implies "conversational". The later is not that easy to implement and is rather specific to each external controller protocol, application and model, etc. Again it is NOT the fit nor the purpose of qtractor's "feedback" function, although it might on very special cases and may require special modification or re-configuration at the external device end-point.

cheers

**Comment by dehnhardt** [comment]
Dear Rui,
2.i works perfect now, **thanks for the quick fix**!
2.ii here I'm not sure if there is a misunderstanding: Currently there in no feedback from qtractor at all when the state of a button is changed from the controller. But when I start qtractor, the initial states of the buttons and faders are send to the controller and the controller is perfectly in sync with qtractor. I expected that kind of feedback as well, when hitting on a controller button. However, as this is working perfectly, when I enable the feedback like i did in this [commit]( https://github.com/rncbc/qtractor/pull/118/commits/468d11006d99e9d039d281e67e8d996443514f3d) it works perfectly for me. So I can maintain my private patch to keep this enabled. I use TouchDAW to control qtractor from another room, so no feedback is somewhat problematic;-)
**Faders:** I think the problem with the faders is too deeply rooted in the system for me to suggest a patch. But I can of course test it if you change that at some point.
**Pan Pots:** I fear, to implement a flexible solution is really out of scope here. I debugged what you did and saw, that you expect a psitive or negative delta to be send by the controller. But my controller only sends: Rotated clockwise or rotated counter clockwise. That explains why the pan faders in Qtractor are always moved to the rightmost position - no matter in which direction I move them in the controller. I simply will not use them or do as with the feedback function and keep a 'private patch'.
Thanks for your ongoing work, Rui!
Holger

**Comment by rncbc** [comment]
Dear Holger,

re. 2.i  :- great, chalked, ticket closed here ;).

re. 2.ii :- as said (maybe more than once) the qtractor design is such that the "feedback" is ONLY issued when on qtractor's initiative; it ain't, never was, meant as an ackowledge/confirming response whatsoever; in fact, it is as simple as a "mirror" option. As is to say: please don't turn it on if the your external controller device don't interpret incoming messages with the very same semantics as it does outgoing. That also makes a point about this new "Delta" mode, which only applies to the controlled end-point ie. qtractor--momentary for toggling subjects (monitor, rec, mute, solo) and differential for scalar ones (gain, panning)--on which cases the current "feedback" implementation on the [midicontrolx](https://github.com/rncbc/qtractor/tree/midicontrolx) branch is, to put it lightly, _asymmetrical_.

re. Faders :- the issue pertaining to audio track faders might have been already fixed -- see https://github.com/rncbc/qtractor/commit/87387e1 .

re. Pan Pots :- well, let's give it another try, this time based on yours: delta will increment/decrement on decimal steps (-/+0.1) -- see https://github.com/rncbc/qtractor/commit/d0a18a8 .

have to tell that the [midicontrolx](https://github.com/rncbc/qtractor/tree/midicontrolx) tree is close to getting merged into master, any day soon ;)

and i still count on you to see what my eyes don't ;)

cheers

**Comment by rncbc** [comment]
on a side note :)
long ago, i had this thing pumping : [qmidictl](http://qmidictl.sf.net) + [qmidinet](http://qmidinet.sf.net) + [usx2y.qtc](http://www.rncbc.org/datahub/usx2y.qtc) ... you probably know the drill...
not that qmidictl is going to replace touchdaw ever, but it's all been a lot of fun doing things "my way" ;)
cheers

**Comment by dehnhardt** [comment]
Hey Rui,
> and i still count on you to see what my eyes don't ;)

Sorry, I'm performing bad here;-) So much to do currently. I will test the additional changes this evening.

Feedback: 
> As is to say: please don't turn it on if the your external controller device don't interpret incoming messages with the very same semantics as it does outgoing.

Hmm still unsure: The feedback which I turned on in that patch simply transmitted the current value of the control (not the values which were sent by the controller) which works perfectly in my case (and of course that should be, as it is exactly the same that is sent when the initial values are transmitted). 
But to come to an end here: I don't think I can harm anyone but me, when I enable it in a 'private patch' - do I;-)

Holger

**Comment by rncbc** [comment]
hi Holger,

> The feedback which I turned on in that patch simply transmitted the current value of the control (not the values which were sent by the controller) which works perfectly in my case (and of course that should be, as it is exactly the same that is sent when the initial values are transmitted).
> But to come to an end here: I don't think I can harm anyone but me, when I enable it in a 'private patch' - do I;-)

Of course you can have "Delta" and "Feedback" simultaneously set, just that the normal direction messages (controller -> qtractor) and the feedback counterpart (qtractor -> controller) will differ in content on that situations. IOW, the "Delta" mode applies only to the way incoming messages are interpreted, and do not apply to respective outgoing ones (when "Feedback" is on.)

hth.
cheers

**Comment by dehnhardt** [comment]
Hi Rui,

even the delta thing with the pan pots works now. Great and thank you. 
The faders on the audio tracks now stops at 1.0f in the controller when moved above 0db in qtractor. consequentially I can not move the faders completely to the top in qtractor when using the controller, but 0db should be enough in most cases.
From my side the pull request can now be closed.

Thanks again
Holger

(BTW: I will test the qmidictl + qmidinet + usx2y.qtc chain soon. Thanks for the tip. qmidictl was completely out of my focus...)

**Comment by rncbc** [comment]
hi,

former [midicontrolx](https://github.com/rncbc/qtractor/tree/midicontrolx) branch is now merged to master HEAD https://github.com/rncbc/qtractor/commit/67b32a0 .

this PR will now put to a close.
thanks a lot.
cheers

**Comment by schnitzeltony** [comment]
@rncbc: Slightly off-topic but can you please merge master to neonx?

**Comment by rncbc** [comment]
@schnitzeltony: done. https://github.com/rncbc/qtractor/commit/688a332
